### PR TITLE
A parametricity-style algorithm for automatic generation of Boolean equality associated to an inductive type

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** Enhancement

Equations, Elpi and the `Scheme` command are known to automatically build schemes associated to inductive types.

Originally motivated by adding support for local definitions in the parameters of inductive types on which `Scheme Equality` applies (!), I realized in passing that the generation of Boolean equality could be both simplified and generalized by following the pattern of parametricity translations for the Calculus of Inductive Constructions (Keller-Lasson, ...). This is what led to this PR.

I was pointed in the meantime to Elpi which is also able to build Boolean equality schemes. I believe it is strongly worth to eventually mix all the knowledge about automatic generation of schemes into a powerful documented system of scheme generation. In the meantime, I'm proposing this PR at least to improve `Scheme Boolean Equality` while something better based e.g. on Elpi subsuming `Scheme Boolean Equality` can be provided.

Technically, compared to the previous version, this PR supports:
- higher-order parameters: e.g. `Inductive I (F:nat->(Type->Type)->Type) := ...`
- non-Type parameters: e.g. `Inductive I (f:nat->nat) := ...`
- local definitions in parameters and constructors
- application of inductive types to non-variables: e.g. `Inductive I := C : list nat -> I`
- not generating equality on `SProp` parameters
- universe polymorphism (thanks to #15525)
- recursively non-uniform parameters                                                                                                                       
- reasoning under a `match`

It still does not support:
- indices (as in `Vector.t`) and dependencies in constructors (as in `sigT`)
   - for `Vector.t`, we could do it by using a generalization in pattern-matching
   - for `sigT`, this would a priori require to manipulate heterogeneous Boolean equality (quite easy to do though)
- parameters encapsulated in data types: e.g. `Inductive I (T:Type*Type) := ...`
- application of sorts, as e.g. in `Inductive I (F:forall A, A -> A) (G:F Type Type) := ...` which could in theory be associated to `I_beq (F_eq:forall A, (A->A->bool) -> (A->A->bool)) (F_eq_type:forall A, (A->A->Type) -> (A->A->Type)) (G:F Type Type) (G_eq:F_eq_type Type (fun A B => A -> B -> bool))`
- reasoning under a `fix` or under projections                                                                                                    

Incidentally, it fixes an issue with internal name collision when using variables `X` and `Y` in types.

Note that this is only for Boolean equality. The algorithms building schemes connecting Boolean equality and Leibniz equality, or proving decidability of equality are unchanged by the PR.

Depends on 
- #15526 
- #15525.

- [x] Added / updated **test-suite**.
- [x] Added **changelog**.

--------

**A few details on the algorithm**

The Boolean equality translation is a parametricity translation where each type `T` is interpreted as the pair of:
- a (possibly degenerate) predicate `T_R` in `Type` over `T` (i.e. `T_R : T->Type`)
- when `T` is decidable, a Boolean equality `T_E` over (i.e. `T_E : option (T->T->bool)` where `None` means not decidable, that is, at worst, unknown to be decidable)

This generalizes into an interpretation of each term `M:T` as:
- an inhabitant `[|M|]` of `T_R M`

by setting for sort s:
- `s_R := \T:s.{R:T->s ; E:option (T->T->bool)}` and
- `s_E := None`

so that types `T:s` are indeed interpreted as a pair, namely by:
- `[|T|] := {R:=T_R; E:=T_E} : s_R T`

and, in particular:
- `[|Type(n)|] := {R:=Type(n)_R; E:=Type(n)_E) : Type(n+1)_R Type(n)`

In practise, to have simpler schemes, we won't support here the full hierarchy of sorts. That is, we assume that type parameters of a type will be instantiated only by small types (i.e. not containing sorts). This makes sense since equality on large types is anyway not decidable in the general case [*].

Now, it happens that several parts of the translation are degenerate. For instance, if `T` is a small type (not containing sorts), then `T_R M`, which expresses that `M` realizes `T`, is a singleton for all `M` (we assume functional extensionality to treat the case of dependent product types). Therefore, `M_T` does not need to be defined in this case. Conversely, if `T` is not small, it is not decidable and `T_E` will be `None`.

This means that at least `T_R` is degenerate or `T_E` is `None` and this also means that for `M:T` with degenerate `T_R`, we don't need to compute `[|M|].R`

This further means that when translating a variable `x:T` with `T` small, it can just be translated to `x:T` without requiring the trivial information that `T_R x` is inhabited.

Similarly for types `T` of the form `forall X, (forall Y ...) ... Y args` based on the assumption [*] that `Y` is instantiated only by small functorial type for which `(Y args)_R` would be degenerate.

Similarly, based on the restriction above [*] that parameters are instantiated by small types, when translating a variable `x:T` with `T` an arity, i.e. of the form `... -> Type` (or at worst `... -> Type*Type` etc.), we can assume `x` to be instantiated by a small functorial type whose `R` translation is degenerate.

It remains the declarations of the form `X:T` for `T` of the form `forall X, ... X args]`where `X` is in negative position of a dependent product. If such `X` is instantiated by a large type in the definition, as in `Inductive I (F:forall X:Type, X->X) (B:F Type nat) := C : B -> I F B`, we give up. Otherwise said, we support only the case when `X`is instantiated by a small type.

Eventually, we thus need two translations:
- `T_R` when `T` is large and it does not need to go under applicative subterms since subterms `X args` are considered small by assumption [*]
- `T_E` when `T` is small which needs to go under subterms and which thus needs to be generalized into a translation `M_E : T_R.E` whenever `T` is detected as decidable (in which case, `T_R.E` is typically of the form `T->T->bool` or of the form `(T.1->T.1->bool)*(T.2->T.2->bool)`, etc.)

In the PR, the first translation is called `translate_type_eq` and the second `translate_term_eq`. Additionally, there is a copy-cat translation called `translate_term` that lifts `M` in the typing context of `M_R`.

The function `translate_type_eq` takes as input a type `T` and a term `M` of type `T` and it returns `T_R M`.

For `M` of type `T`, the function `translate_term_eq M` returns an object of type term `T_R M`, that is, when `M` is itself some type `U`, an object of type `U->U->bool`.

Note that we don't use an option type for non-decidable types but instead raises an exception whenever a type is not detected as decidable.

Future work might support:
- exotic types such as `Inductive I (F:forall X:Type, X->X) (B:F Type nat) := C : B -> I F B`
- types with invertible indices like `Vector.t`
- dependent products over finite types (e.g. over `bool->bool`), and more generally compact types whose equality is decidable (see Escardó-Oliva)

At first view, the algorithm is also reusable for proving the "lb" and "bl" lemmas `x = y -> eqb x y = true`, `eqb x y = true -> x = y` which `Scheme Equality` also proves. For that, we would need to respectively replace `T_E` with:
- `option (forall x y : T, x = y -> eqb x y = true)`
- `option (forall x y : T, eqb x y = true -> x = y)`

and to respectively replace the use of `andb (f x1 y1) ... (f xn yn)` in the definition of `Ind_R.E` by proofs
- `fun H : C x1..xn = C y1..yn => f_eqb (f x1 y1 (inj1 H)) .. (f xn yn (injn H))`
- `fun H : andb (eqb x1 y1) ... (eqb xn yn) = true => f_equal (f x1 y1 (inj1_andb1 H)) .. (f xn yn (injn_andbn H))`

where `f` is the recursive call and `f_equal`, `f_eqb`, `inji`, `inj_andbi` are appropriate congruence and injectivity lemmas.

It could also be used to (directly) prove `{x = y} + {x <> y}`, without actually requiring the "lb" and "bl" lemmas.

### CODE REVIEW

File: test-suite/success/SchemeEquality.v
Comment (JasonGross) [inline test-suite/success/SchemeEquality.v:43]: Should there also be a test for an inductive named `X`?

File: test-suite/success/SchemeEquality.v
Comment (JasonGross) [inline test-suite/success/SchemeEquality.v:97]: Should we check the body of the resulting boolean equality to ensure it was correctly generated?

File: test-suite/success/SchemeEquality.v
Comment (herbelin) [reply test-suite/success/SchemeEquality.v:43]: I don't think so. I added the test to show that the bug existed before, but the new implementation does not rely on the name of variables anymore, so I don't think the test would really add something.

File: test-suite/success/SchemeEquality.v
Comment (herbelin) [reply test-suite/success/SchemeEquality.v:97]: That'll be nice. We'll do.

**Comment by JasonGross** [comment]
> * `s_R := \T:s.{R:T->s ; E:option (T->T->s)}` and

Should this be `E:option (T -> T -> bool)`?

Also, would it make sense to include your explanation of this translation in the documentation of scheme bookman equality?

**Comment by herbelin** [comment]
> > * `s_R := \T:s.{R:T->s ; E:option (T->T->s)}` and
> 
> Should this be `E:option (T -> T -> bool)`?

Indeed, thanks. Fixed.

> Also, would it make sense to include your explanation of this translation in the documentation of scheme bookman equality?

You mean the reference manual? The explanation is a bit long, but maybe can it be shortened, or put in an annex. I don't know what others would think. If not already known in the literature, it should ideally be a paper. I don't know if co-authors would be interested in the topic.

**Comment by herbelin** [comment]
There is room for various extensions in this PR but I would need help to go further. It is already a valuable first step I think and I'm releasing it for review.

File: kernel/constr.ml
Comment (SkySkimmer) [inline kernel/constr.ml:853]: Is ths change actually useful?

Comment (herbelin) [reply kernel/constr.ml:853]: Useful for the PR no. Useful in general I even don't know since the mkRel are already mostly all shared. It was a change in passing that I did not remove (will remove it).

Similarly, the `esubst.mli` documentation was done in passing and is not directly related to the PR.

**Comment by gares** [comment]
Can you add a test with rose trees?
```ocaml
type 'a rose := Leaf of 'a | Node of  ('a rose) list
```

File: kernel/constr.ml
Comment (SkySkimmer) [reply kernel/constr.ml:853]: please drop this code complexification

**Comment by herbelin** [comment]
> > It would be nice if you could add all of this, actually: https://github.com/LPCIC/coq-elpi/blob/master/apps/derive/tests/test_derive_stdlib.v

Good idea. Elpi test-suite now added.

 I removed `Bush` (not supported by Coq), `Int63` and `Float` (no primitive test), `w` and `dyn` (not decidable in general), `vect` and `dep_record` (would need more work, see hints above), `iota` (dependencies).

It revealed a bug with using the compatibility level for primitive projections and I seized the opportunity to produce directly the expected term (without `let` expansions). It also revealed a badly-caught error message which I now fixed, thanks.

File: vernac/auto_ind_decl.ml
Comment (jfehrle) [inline vernac/auto_ind_decl.ml:328]:
```suggestion
     In practice, to have simpler schemes, we won't support here the
```

File: vernac/auto_ind_decl.ml
Comment (jfehrle) [inline vernac/auto_ind_decl.ml:211]:
```suggestion
    [eq_status] tells how to get the equality associated with a variable if any
```

Same change in the title, too

**Review by jfehrle — COMMENTED** [review]
What's the status of the 6 "TODO"s?  Are they needed now for correctness or are they potential future enhancements?

Also, if this is a user-visible change, shouldn't there be a changelog entry and/or documentation?

File: vernac/auto_ind_decl.ml
Comment (herbelin) [reply vernac/auto_ind_decl.ml:328]: Thanks.

**Comment by herbelin** [comment]
> What's the status of the 6 "TODO"s? Are they needed now for correctness or are they potential future enhancements?

I now updated two TODOs which were obsolete.

It remains two pairs of TODOs that refer to potential enhancements and that are not needed for correctness (I also added one of such):
- completing the Fix case
- thinking to a way to extend to Type arguments encapsulated in some Ind (and inhabitants obtained with Construct/CoFix/Proj)
- completing the Array case (should we iterate using `PArray.length` and `PArray.get`??)

Note incidentally that I just completed the cases for Int63 and Float (@gares and Elpi:  for that purpose, I exported Coqlib names `num.int63.eqb` and `num.float.eqb` directly from the original files if ever you want to reuse them).

More significantly, there is the question of plug the so-called `lb` and `bl` schemes in the new algorithm; for that, one would typically need first the so-called `NoConfusion` scheme; there is such a NoConfusion algorithm hidden in file `equality.ml` but Equations and Elpi already have the infrastructure to bind the NoConfusion scheme to constants, so, we could first find a way to integrate Equations and Elpi's scheme mechanism in Coq.

> Also, if this is a user-visible change, shouldn't there be a changelog entry and/or documentation?

There should eventually be a changelog. The current status is to get feedback on the extent to give to the PR and about the plans we can make about producing schemes in general.

File: theories/Floats/PrimFloat.v
Comment (herbelin) [reply theories/Floats/PrimFloat.v:53]: I know that equality on floats is not reflexive (which is a bit strange actually), and, indeed, it is questionable whether we should qualify such relation as an equality.

In practice, nothing requires a Boolean equality passed as argument of a Boolean equality scheme to be reflexive, so this is not worse, say, than calling `list_beq` on `Float.eqb`. That is, we can consider this has a "you're supposed to know what you're doing when using equality of float".

But maybe should we instead refuse to build a Boolean equality for types mentioning float, I don't know.

**Comment by herbelin** [comment]
There is a lot to do more in the direction of this PR but I'm stopping there.
- I tried to add the `fix` case but failed. I would need help from parametricity specialists (can we do it without bundling a type `A` with its equality `eq_A`? I have interferences between the translation of `match` and `fix`)
- it should be extended to `bl` and `lb` schemes, then further to `eqdec` schemes (up to make the latter compositional rather than dependent on `bl` and `lb`)
- a more general reflection on schemes should be done: what general mechanism? how to support any constant and not only inductive types? how to integrate Elpi and Equations schemes?

> Also, would it make sense to include your explanation of this translation in the documentation of scheme bookman equality?

I don't know what refman maintainers think about it.

> Note that this is the IEEE754 equality and not the boolean equality 

I would say it is ok but maybe worth collecting other opinions.

**Review by Alizter — APPROVED** [review]

**Comment by Alizter** [comment]
Will merge tomorrow when CI is green.

**Comment by Alizter** [comment]
Hmm there seems to be some issues with fiat_crypto. Since I am suspecting things are taking longer, perhaps a bench is in order?

@SkySkimmer Can you do a bench?

**Comment by gares** [comment]
@Alizter don't you gave the rights to push on the gitlab button in the pipeline view?

**Comment by Alizter** [comment]
@gares Doesn't look like it.

**Comment by herbelin** [comment]
Iirc, fiat-crypto timeout several times in the PR, so there may be specific problem there.

Technically, the algorithm is linear in the size of the inductive declaration + a few whd-reductions to eta-expand terms to compensate that signatures have the form A, B, A_eq, B_eq rather than A, A_eq, B, B_eq. The algorithm is also supposed to succeed more often.

Maybe fiat-crypto has a complex inductive declaration with a long whd-reduction? I see occurrences of `Set Boolean Equality Type` in 9 files.

Is the version of fiat-crypto in the bench the same as in the CI?

**Comment by JasonGross** [comment]
> Maybe fiat-crypto has a complex inductive declaration with a long whd-reduction? I see occurrences of `Set Boolean Equality Type` in 9 files.

We do emit inductives with dozens or perhaps hundreds of constructors, and jump through some hoops to build efficient equality for them.  What files is it timing out in / what's the performance of various files?

**Comment by herbelin** [comment]
> We do emit inductives with dozens or perhaps hundreds of constructors, and jump through some hoops to build efficient equality for them. What files is it timing out in / what's the performance of various files?

Fiat-crypto is stopped by the 3h CI timeout. Last file it was compiling is `src/Curves/EdwardsMontgomery.v`. Bench (see in CI) does not show extra time. I don't know which file is timing out.

**Comment by CohenCyril** [comment]
I have several questions about this PR:
- are you reimplementing a parametricity translation ?
- is the scheme command here subsumed by another one (Equation's or coq-elpi's) ?

**Comment by herbelin** [comment]
> I have several questions about this PR:
> 
>     * are you reimplementing a parametricity translation ?

Yes, but "optimized" to the objective of producing Boolean equality (see explanations in the header).

>     * is the scheme command here subsumed by another one (Equation's or coq-elpi's) ?

This was my question at some time. I was directed to elpi derive which I discovered with a very big enthousiasm as it is many years that I'm dreaming of a general toolbox for generating a variety of schemes. (By the way, such work should not remain isolated, it should be made more widely usable so that others can rely on it and build new schemes further on it. I know it is written in a Prolog, but I'm pretty sure that with MetaCoq techniques, the schemes could be produced also under the form of a `constr`.)

So, I tried elpi derive which was fantastic. It is possible that I did not know how to do it properly, but there were some Boolean equalities that I was not able to produce, e.g. in the presence of higher-order arities, or in the presence of `if` in the type. I'm sure it could be extended though.

Wrt the current mechanism, note that I aimed at compatibility which forced to some oddities, like `prod` taking `A`, `B`, `A_eq` and `B_eq` rather than `A`, `A_eq`, `B` and `B_eq` which would be more compositional. So, at some time, we would have to revise anyway the mechanism to be "cleaner" and more extensible. That will be a good opportunity to do it with elpi.

**Comment by Alizter** [comment]
I had a closer look at the bench results, there are no performance issues. What I don't understand is that the GitLab job exited with signal 9 which means a kill signal was sent. It also was only 107 minutes and not within the 3h timeout...

**Comment by jfehrle** [comment]
How about rerunning the bench, then?

**Comment by SkySkimmer** [comment]
If we compare failing job https://gitlab.com/coq/coq/-/jobs/2082281764
>src/Assembly/WithBedrock/Semantics.vo (real: 1922.52, user: 1210.92, sys: 7.72, mem: 3539320 ko)

recent master https://gitlab.com/coq/coq/-/jobs/2087378278
>src/Assembly/WithBedrock/Semantics.vo (real: 1.94, user: 0.69, sys: 0.25, mem: 501760 ko)

quite a difference

**Comment by SkySkimmer** [comment]
This file doesn't seem to be in the bench, probably because it does `without-bedrock` https://github.com/coq/opam-coq-archive/blob/9e6513b88fdb69d735e0c35cc253b7c62b9c8a8e/extra-dev/packages/coq-fiat-crypto/coq-fiat-crypto.dev/opam#L14

**Comment by herbelin** [comment]
Thanks for the diagnosis pointing to the file `Semantics.v`. It was about a record with a huge definition and the function to compute the needed dependencies became too naively inefficient. I fixed it.

The `Semantics.v` example (called `machine_state`) is otherwise pretty interesting. It refers to a record with two fields in `Type` which are later projected, that is, it is based (indirectly) on a record of the form `Record T (A:Type*Type) := {a1 : fst A; a2 : snd A}`. This currently goes out of the scope of what the algorithm supports as it enters the category of types with parameters encapsulated in data types.

Note that the Boolean Equality was not generated before either.

Additionally, I changed the algorithm so that it reduces defined constants (as before) rather than failing. This is not optimal as one would prefer that Boolean equalities are attached to constants too, but it has the advantage to be closer to the previous behavior.

I incidentally wondered whether equality on Prop shouldnt' be consider trivial, like it is a in SProp.  This would be useful to generate the equality for `machine_state` which precisely has some fields of type `eq`.

In any case, rerunning the bench.

**Comment by Alizter** [comment]
@herbelin Did you mean to push some commits before rerunning the bench?

**Comment by JasonGross** [comment]
> I incidentally wondered whether equality on Prop shouldnt' be consider trivial, like it is a in SProp. This would be useful to generate the equality for `machine_state` which precisely has some fields of type `eq`.

I think yes.  There's no possibility of doing anything else in a computational way.  The proof of correctness will have to parameterize over UIP for Props which cannot be automatically proven singleton.

**Comment by herbelin** [comment]
Renouncing, at least temporarily, to discard subterms in `Prop` in the Boolean scheme, since proving the equivalence with propositional equality would then require a proof-irrelevance axiom.

File: theories/Floats/PrimFloat.v
Comment (proux01) [reply theories/Floats/PrimFloat.v:54]: The shortest test for nan is `PrimFloat.is_nan` (which is just `fun x => ~~(x =? x)`).
The equality seem currently missing. If it doesn't need to be lightning fast (i.e. if we accept going through `positive` and `Z`), the easier solution may be to use `Prim2SF` and compare the resulting inductive `SpecFloat.spec_float` (free equality with the scheme I guess). If it has to be really  fast, we should export the OCaml implem as a primitive function.

In any case, it may be good to add a comment to `PromFloat.eqb` stating it is the IEEE754 equality (with +0 = -0 and nan = nan being false) and not a regular boolean equality). Would you mind doing it here, or do you prefer me to do it in a separate PR?

Comment (herbelin) [reply theories/Floats/PrimFloat.v:54]: > The shortest test for nan is `PrimFloat.is_nan` (which is just `fun x => ~~(x =? x)`).

So obvious now that you say it! But then, `negb (x=?x) && negb (y=?y) || (x=?y)` should do the trick no? Or should we expect +0 to be distinct from -0?

> In any case, it may be good to add a comment to `PromFloat.eqb` stating it is the IEEE754 equality (with +0 = -0 and nan = nan being false) and not a regular boolean equality). Would you mind doing it here, or do you prefer me to do it in a separate PR?

The comment I had before is still there but I can remove it and let the phrasing to you (or you can also make suggestions on the PR). Currently, the comment talks only about nan (since +0 = -0 is valid in terms of an equality).

Comment (herbelin) [reply theories/Floats/PrimFloat.v:54]: > Or should we expect +0 to be distinct from -0?

The question is I guess: is there a way to observe that `+0` is different from `-0`?

Comment (proux01) [reply theories/Floats/PrimFloat.v:55]: Well, the proof will be an axiom since it will be a primitive operation (just as the soundness of any other primitive float operator).

**Comment by Alizter** [comment]
OK merging this soon.

**Comment by Alizter** [comment]
@coqbot merge now

**Comment by proux01** [comment]
@Alizter please announce it a few day before, when merging such non trivial PRs!
The floating-point point above ( https://github.com/coq/coq/pull/15527#discussion_r797673457 ) was not yet resolved (we should have waited a few more days).
I'll open a separate PR.

**Comment by Alizter** [comment]
@proux01 I had originally announced my intent to merge 9 days ago, but we hit a snag with the CI. I missed the further discussion on the floating point issues due to being buried further in the timeline in some comments. I'm not sure what I could have done differently, I simply did not see there were pending issues.

**Comment by proux01** [comment]
I agree some active discussions can be deeply burried in the thread and don't have any magic solution for that.

At least when a PR is still active (Hugo was still pushing to the branch a few hours before you merged and did not express any intent to stop doing so), don't say "I'm merging now" but "I'll merge tomorrow". I was in the middle of some orchestra rehersal between you "mrging soon" message and the merge, no way I could have expressed anything, whereas this would have been possible in a one day period.

**Comment by Alizter** [comment]
Anyway sorry about that, will do better next time.

**Comment by proux01** [comment]
No worries, I'll just fix that in a separate PR.

**Comment by JasonGross** [comment]
> magic solution for that.

My magic solution for this is subscribing to issues and checking my inbox.  This has the downside of sending me many emails

**Pull request merged**
